platform/cog-platform-fdo.c

```diff
@@ -1664,6 +1689,139 @@ on_export_fdo_egl_image(void *data, struct wpe_fdo_egl_exported_image *image)
+    wl_shm_pool_destroy (buffer->shm_pool);
+    munmap (buffer->data, buffer->size);
+
+    g_free (buffer);
+}
+
+static void
+shm_buffer_destroy_notify (struct wl_listener *listener, void *data)
+{
+    struct shm_buffer *buffer = wl_container_of (listener, buffer, destroy_listener);
+
+    wl_list_remove (&buffer->link);
+    shm_buffer_destroy (buffer);
+}
+
+static void
+on_shm_buffer_release (void *data, struct wl_buffer *wl_buffer)
+{
+    struct shm_buffer* buffer = data;
+    if (buffer->exported_buffer) {
+        wpe_view_backend_exportable_fdo_egl_dispatch_release_shm_exported_buffer (wpe_host_data.exportable,
+                                                                                  buffer->exported_buffer);
+        buffer->exported_buffer = NULL;
+    }
+}
+
+static const struct wl_buffer_listener shm_buffer_listener = {
+    .release = on_shm_buffer_release,
+};
+
+static void
+shm_buffer_copy_contents (struct shm_buffer *buffer, struct wl_shm_buffer *exported_shm_buffer)
+{
+    int32_t height = wl_shm_buffer_get_height (exported_shm_buffer);
+    int32_t stride = wl_shm_buffer_get_stride (exported_shm_buffer);
+
+    size_t data_size = height * stride;
+
+    wl_shm_buffer_begin_access (exported_shm_buffer);
+    void* exported_data = wl_shm_buffer_get_data (exported_shm_buffer);
+
+    memcpy (buffer->data, exported_data, data_size);
+
+    wl_shm_buffer_end_access (exported_shm_buffer);
+}
+
+static void
+on_export_shm_buffer(void* data, struct wpe_fdo_shm_exported_buffer* exported_buffer)
+{
+    struct wl_resource *exported_resource = wpe_fdo_shm_exported_buffer_get_resource (exported_buffer);
+    struct wl_shm_buffer *exported_shm_buffer = wpe_fdo_shm_exported_buffer_get_shm_buffer (exported_buffer);
+
+    struct shm_buffer *buffer = shm_buffer_for_resource (exported_resource);
+    if (!buffer) {
+        int32_t width = wl_shm_buffer_get_width (exported_shm_buffer);
+        int32_t height = wl_shm_buffer_get_height (exported_shm_buffer);
+        int32_t stride = wl_shm_buffer_get_stride (exported_shm_buffer);
+        uint32_t format = wl_shm_buffer_get_format (exported_shm_buffer);
+
+        size_t size = stride * height;
+        buffer = shm_buffer_create (exported_resource, size);
+        if (!buffer)
+            return;
+        wl_list_insert (&wl_data.shm_buffer_list, &buffer->link);
+
+        buffer->buffer = wl_shm_pool_create_buffer (buffer->shm_pool, 0, width, height, stride, format);
+        wl_buffer_add_listener (buffer->buffer, &shm_buffer_listener, buffer);
+    }
+
+    buffer->exported_buffer = exported_buffer;
+    shm_buffer_copy_contents (buffer, exported_shm_buffer);
+
+    wl_surface_attach (win_data.wl_surface, buffer->buffer, 0, 0);
+    wl_surface_damage (win_data.wl_surface,
+                       0, 0,
+                       win_data.width * wl_data.current_output.scale,
+                       win_data.height * wl_data.current_output.scale);
```

Comment:
Recently I have seen code that passes `INT32_MAX` as width and height when the goal is to invalidate a surface: the damaged area is clipped to the dimensions of the surface anyway. 

It's not like it should make much of a difference in modern hardware to save two multiplications, but dunno, in my head it looks clearer that seeing `0, 0, INT32_MAX, INT32_MAX` the call means “invalidate the whole thing”—but I am not sure if this is only my impression, so feel free to leave it as-is.